Add context menu for multi-file entries (#12567)

w0nderfu11 opened the pull request:
Extend the Entry Editor context menu to handle entries with multiple linked files. This improves UX when managing more than one file per entry and aligns behavior across single- and multi-file scenarios.

Changes:
- Add plural actions to StandardActions enum
- Rewrite ContextAction.execute() for multi-file cases
- Extend ContextMenuFactory to build multi-file items
- Rework MultiContextAction to operate on selections
- Introduce CopyMultipleFilesAction (new class)
- Update/add localization keys; tests pass
- Add unit tests: ContextActionTest, ContextMenuFactoryTest, MultiContextActionTest, CopyMultipleFilesActionTest
- Guard LinkedFileViewModel to avoid a JavaFX crash

Fixes #12567
Keywords: context menu, linked files, multi-selection, UI

Closes _____
<!-- LINK THE ISSUE WITH THE "Closes" KEYWORD. Example: Closes https://github.com/JabRef/jabref/issues/13109 OR Closes #13109 -->

<!-- In about one to three sentences, describe the changes you have made: what, where, why, ... (REPLACE THIS LINE) -->

<!-- NOTE: If your work is not yet complete, please open a draft pull request. In that case, outline your intended next steps. Do you need feedback? Will you continue in parallel? ... -->

### Steps to test

<!-- Describe how reviewers can test this fix/feature. Ideally, think of how you would guide a beginner user of Jabef to try out your change. -->
<!-- You can add screenshots or videos (using Loom - https://www.loom.com or by just adding .mp4 files). -->
<!-- (REPLACE THIS PARAGRAPH) -->

<!-- YOU HAVE TO MODIFY THE ABOVE TEXT FIT YOUR PR. OTHERWISE, YOUR PR WILL BE CLOSED WITHOUT FURTHER COMMENT. -->

### Mandatory checks

<!--
Go through the checklist below. It is mandatory, even for a draft pull request.

Keep ALL the items. Replace the dots inside [.] and mark them as follows: 
[x] done 
[ ] TODO (yet to be done)
[/] not applicable
-->

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [.] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [.] Tests created for changes (if applicable)
- [.] Manually tested changed features in running JabRef (always required)
- [.] Screenshots added in PR description (if change is visible to the user)
- [.] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [.] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

jabref-machine commented:
Note that your PR will not be reviewed/accepted until you have gone through the mandatory checks in the description and marked each of them them exactly in the format of `[x]` (done), `[ ]` (not done yet) or `[/]` (not applicable).


<!-- ghprcomment
PR Tests
PR checklist OK
-->

jabref-machine commented:
Note that your PR will not be reviewed/accepted until you have gone through the mandatory checks in the description and marked each of them them exactly in the format of `[x]` (done), `[ ]` (not done yet) or `[/]` (not applicable).


<!-- ghprcomment
PR Tests
PR checklist OK
-->

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/copyfiles/CopyMultipleFilesAction.java:
```diff
@@ -0,0 +1,89 @@
+            Path dst = targetDir.resolve(src.getFileName());
+
+            if (Files.exists(dst) && Files.isDirectory(dst)) {
+                IOException ex = new IOException("Destination is a directory: " + dst);
```
In this case, the `IOException` is not used for control flow — 
we explicitly check the condition first, show an error dialog, and then skip the file with `continue`.  
Using the exception object here is only for passing context to the UI, so I believe this approach is appropriate.

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/copyfiles/CopyMultipleFilesAction.java:
```diff
@@ -0,0 +1,89 @@
+import org.jabref.logic.l10n.Localization;
+import org.jabref.model.database.BibDatabaseContext;
+
+public class CopyMultipleFilesAction extends SimpleCommand {
```
If needed can try to separate it and not broke some alredy done fucntionality by me

w0nderfu11 commented:
Hi @koppor, all checks have passed ✅ Could you please review and approve when you have time?  
Thanks a lot!

subhramit commented:
> i set up checkstyle with your guide but it doesn`t show me some problems in test classes, only in main module)

<img width="1934" height="1467" alt="image" src="https://github.com/user-attachments/assets/f2dc710b-9476-4814-9e4e-4276c9365c3d" />

Check this setting (taken from [here](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#put-jabrefs-checkstyle-configuration-in-place))

calixtus commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/MultiContextAction.java:
```diff
@@ -39,16 +47,136 @@ public MultiContextAction(StandardActions command,
+        }
+    }
+
+    private boolean isLocalAndExists(LinkedFileViewModel vm) {
+        return !vm.getFile().isOnlineLink()
+                && vm.getFile().findIn(databaseContext, preferences.getFilePreferences()).isPresent();
+    }
+
+    private boolean isOnline(LinkedFileViewModel vm) {
+        return vm.getFile().isOnlineLink();
+    }
+
+    private boolean hasSourceUrl(LinkedFileViewModel vm) {
+        return !vm.getFile().getSourceUrl().isEmpty();
+    }
+
+    private boolean isMovableToDefaultDir(LinkedFileViewModel vm) {
+        return isLocalAndExists(vm) && !vm.isGeneratedPathSameAsOriginal();
+    }
```
Can those predicates be reused in the (SingleFile)ContextFactory? If yes, package private. In that case, tests will be way easier.

w0nderfu11 replied:
This class was deleted cause of restructing logic, but i created an interface with such "helpers" methods and in some points used existing logic with executable or no

calixtus commented on jabgui/src/test/java/org/jabref/gui/copyfiles/CopyMultipleFilesActionTest.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    @Test
+    void copiesAllSelectedFilesToChosenDirectory() {
```
TempDir can be created for test method only
```suggestion
    void copiesAllSelectedFilesToChosenDirectory(@TempDir Path tmpDir) {
```

calixtus commented on jabgui/src/test/java/org/jabref/gui/copyfiles/CopyMultipleFilesActionTest.java:
```diff
@@ -0,0 +1,194 @@
+
+class CopyMultipleFilesActionTest {
+
+    @TempDir
+    Path tmpDir;
```
method-specific

calixtus commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/ContextMenuFactoryTest.java:
```diff
@@ -21,54 +27,54 @@
-                // Toolkit already initialized by another thread/test
-            }
-            toolkitInitialized = true;
+    static void initToolkit() throws InterruptedException {
+        try {
+            Platform.startup(() -> { });
+        } catch (IllegalStateException alreadyStarted) {
+            CountDownLatch latch = new CountDownLatch(1);
+            assertDoesNotThrow(() -> Platform.runLater(latch::countDown));
+            assertTrue(latch.await(1, TimeUnit.SECONDS), "JavaFX Platform should be running");
         }
     }
```
I think this might be avoidable with `@ExtendWith(ApplicationExtension.class)`

calixtus reviewed (changes requested):
Thanks for your PR so far.
Some remarks made.

I noticed the Strategy Pattern was somewhat broken there (not by you). Please fix it.
<img width="1116" height="271" alt="grafik" src="https://github.com/user-attachments/assets/85a6767d-b5b8-4882-9607-82ea5fb5ba28" />

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/copyfiles/CopyMultipleFilesAction.java:
```diff
@@ -0,0 +1,89 @@
+    private final BibDatabaseContext databaseContext;
+    private final FilePreferences filePreferences;
+
+    public CopyMultipleFilesAction(ObservableList<LinkedFileViewModel> selectedFiles,
```
removed CopyMultipleFilesAction and extended CopySingleFileAction to accept a list of files. Multi-selection now calls the same action with multiple inputs; single-selection still passes one file. This reduces duplication without changing behavior.

Will see it on future PR (maybe draft)

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/actions/StandardActions.java:
```diff
@@ -106,7 +106,9 @@ public enum StandardActions implements Action {
 
     OPEN_DATABASE_FOLDER(Localization.lang("Reveal in file explorer")),
     OPEN_FOLDER(Localization.lang("Open folder"), Localization.lang("Open folder"), IconTheme.JabRefIcons.FOLDER, KeyBinding.OPEN_FOLDER),
+    OPEN_FOLDERS(Localization.lang("Open folders"), Localization.lang("Open folders"), IconTheme.JabRefIcons.FOLDER, KeyBinding.OPEN_FOLDER),
```
I deleted additional commands, using only existing (except REMOVE_LINKS, it was implemented not by me so decided to let it go), fixed enum and localisation, implemented with Strategy pattern also

subhramit commented:
> CopyMultipleFilesAction and extended CopySingleFileAction to accept a list of files. Multi-selection now calls the same action with multiple inputs; single-selection still passes one file. This reduces duplication without changing behavior.

> Thanks for the heads-up. I’ve restored the Strategy pattern: ContextMenuFactory now delegates to ContextMenuBuilder strategies (SingleSelectionMenuBuilder / MultiSelectionMenuBuilder) with shared checks in SelectionChecks. This removes the ad-hoc branching/factories, and fixes the NPE (null observable) while keeping the same user behavior and proper disabled states.

1. Please refrain from using [AI to communicate here](https://github.com/JabRef/jabref/blob/main/CONTRIBUTING.md#notes-on-ai-usage).
2. Do not resolve conversations before pushing or actually changing the state on GitHub. There can be follow-uo discussions on existing threads, or there could also be errors.
3. Your IntelliJ screenshots don't show anything valuable, so you can skip pasting them in comments unless there is any substance to show.
4. Do not plan changes related to this PR in a follow-up. Address comments corresponding to this PR in this open PR itself.

w0nderfu11 commented:
> > CopyMultipleFilesAction and extended CopySingleFileAction to accept a list of files. Multi-selection now calls the same action with multiple inputs; single-selection still passes one file. This reduces duplication without changing behavior.
> 
> > Thanks for the heads-up. I’ve restored the Strategy pattern: ContextMenuFactory now delegates to ContextMenuBuilder strategies (SingleSelectionMenuBuilder / MultiSelectionMenuBuilder) with shared checks in SelectionChecks. This removes the ad-hoc branching/factories, and fixes the NPE (null observable) while keeping the same user behavior and proper disabled states.
> 
> 1. Please refrain from using [AI to communicate here](https://github.com/JabRef/jabref/blob/main/CONTRIBUTING.md#notes-on-ai-usage).
> 2. Do not resolve conversations before pushing or actually changing the state on GitHub. There can be follow-up discussions on existing threads, or there could also be errors.
> 3. Your IntelliJ screenshots don't show anything valuable, so you can skip pasting them in comments unless there is any substance to show.
> 4. Do not plan changes related to this PR in a follow-up. Address comments corresponding to this PR in this open PR itself.

1. I'm a little bad at getting the wording right. Will not use it to communicate, ty
2. Okay, my bad.
3. I mean it`s not a prove, it`s just showing of work, thath`s all. If you need only correct push with changes i can do it without additional info.
4. I meant remade this PR (or do this PR as draft, i guess i can convert it back, cause it was before)

w0nderfu11 commented:
Marked issues unresolved, will make push soon for review, thx

jabref-machine commented:
Note that your PR will not be reviewed/accepted until you have gone through the mandatory checks in the description and marked each of them them exactly in the format of `[x]` (done), `[ ]` (not done yet) or `[/]` (not applicable).


<!-- ghprcomment
PR Tests
PR checklist OK
-->

subhramit commented:
> @subhramit how can i restart PR Test? i have only this issue

I can see them passing - which one are you concerned about?

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditorViewModel.java:
```diff
@@ -104,6 +108,39 @@ public static LinkedFile fromFile(Path file, List<Path> fileDirectories, Externa
+            fileViewModel.downloadProgressProperty().addListener(ignore -> recompute.run());
+        }
+
+        files.addListener((javafx.collections.ListChangeListener<LinkedFileViewModel>) change -> {
```
fixed

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/SingleSelectionMenuBuilder.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Override
+    public List<MenuItem> buildMenu(ObservableList<LinkedFileViewModel> selection) {
+        LinkedFileViewModel selectedLinkedFile = selection.getFirst();
```
fixed

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecks.java:
```diff
@@ -0,0 +1,58 @@
+
+        for (Map.Entry<Path, List<Path>> entry : filesByDirectory.entrySet()) {
+            Path fileToSelect = entry.getValue().getFirst();
+            try {
+                NativeDesktop.get().openFolderAndSelectFile(fileToSelect);
+            } catch (IOException ioException) {
```
fixed

w0nderfu11 commented:
@Siedlerchr @subhramit @calixtus PR checklist is OK. I made all what we discussed above: fixed Pattern Strategy, also merged copy logic for multiselected files. Due to this there are many major changes, fully reworked classe etc. 

SO i kindly ask you if it`s possible mark as resolved old coversations and made fixes in new code alredy, if old one is OK.

Cause i`m a bit confused now with amount of conversations :)

w8iting for your review again and ready to fix things to make it done already

Also updated screenshot on PR and description!

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditor.java:
```diff
@@ -184,7 +188,7 @@ private Node createFileDisplay(LinkedFileViewModel linkedFile) {
 
         Node icon = linkedFile.getTypeIcon().getGraphicNode();
-        icon.setOnMouseClicked(event -> linkedFile.open());
+        icon.setOnMouseClicked(ignore -> linkedFile.open());
```
fixed, did everywhere with underscore

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/copyfiles/CopySingleFileAction.java:
```diff
@@ -14,52 +18,107 @@
 import org.jabref.model.entry.LinkedFile;
-import org.jabref.model.util.OptionalUtil;
 
 public class CopySingleFileAction extends SimpleCommand {
 
-    private final LinkedFile linkedFile;
+    private final List<LinkedFile> linkedFiles;
```
Class renamed for CopyLinkedFilesAction, hope it would be OK

calixtus commented:
Sorry for delay, after 4 days of JabCon we have been a bit exhausted. I will look at your changes tonight. I will also try to fix the merge conflicts. We enhanced our automatic style checker, which lead to reformatting about 600 files in the code.

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditorViewModel.java:
```diff
@@ -93,7 +99,6 @@ private static String getStringRepresentation(List<LinkedFileViewModel> files) {
     /**
      * Creates an instance of {@link LinkedFile} based on the given file.
      * We try to guess the file type and relativize the path against the given file directories.
-     *
```
returned the "*" sign

calixtus commented:
About style update we posted in the gutter chat:

⚠️⚠️ We updated the code style and reformatted all code ⚠️⚠️

If you ware working on something, you need to update the JabRef code style in the IDE. (A fresh setup is not affected.)

1. Save and commit all your work.
2. Quit IntelliJ.
3. `git fetch upstream` to ensure that `upstream` is most recent
4. `git checkout main`
5. `git merge upstream/main` to ensure that your `main` branch is up to date.
6. Follow the section at https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-auto-formatting-of-your-code to enable auto format
7. Start IntelliJ
8. File > Invlidate Caches... - and then "Invalidate and Restart (just to ensure that the new code style is really read)
9. Reformat your **modified classes** - otherwise, you will have merge conflicts.
10. Commit the reformat
11. Merge upstream/main

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditor.java:
```diff
@@ -301,36 +301,34 @@ private void addFromURL() {
                 activeContextMenu = null;
             }
+            return;
+        }
+
+        if (event.getButton() == MouseButton.SECONDARY) {
```
fixed

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/LinkedFileViewModel.java:
```diff
@@ -203,8 +203,6 @@ public void acceptAsLinked() {
 
     public Observable[] getObservables() {
         List<Observable> observables = new ArrayList<>(Arrays.asList(linkedFile.getObservables()));
-        observables.add(downloadOngoing);
-        observables.add(downloadProgress);
         observables.add(isAutomaticallyFound);
         return observables.toArray(new Observable[0]);
     }
```
explained in other converstaion (made a mistake, thought i wrtoe here), os all now done

w0nderfu11 commented:
@calixtus bro idk what happened, also i run rewrite task in gradle, i got billion of midifications, all is failling idk why)
I will try to resolve it now but this is so hard to understand, I used your guidline about new rules but anyway

w0nderfu11 commented:
@calixtus i reset hard my branch to commit i did not any fixes, i will try to do all by myself, if i have problems i will tell you, don`t worry for now please, i will be back soon

jabref-machine commented:
JUnit tests of `jablib` are failing. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Unit tests (pull_request)" and click on it.

You can then run these tests in IntelliJ to reproduce the failing tests locally. We offer a quick test running howto in the section [Final build system checks](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-12-build.html#final-build-system-checks) in our setup guide.


<!-- ghprcomment
Source Code Tests
Unit tests – jablib
-->

w0nderfu11 commented:
@calixtus i have only label PR failed check, so others are done. I fixed conflicts.

In this push you can see other files, i simply restore them from upstream/main and did reformat of code, i changed for fact only jabgui module all the things you told me above.

Now you can check it!

I hope there will be no more fixes needed. If what i ready to fix some. Hope all is good now :) (Tired a bit)

calixtus commented:
Was this your last change to the code?

https://github.com/JabRef/jabref/pull/13726/commits/259e37c7372e5067821fdad5125a27c0e572fab0

w0nderfu11 commented:
> Was this your last change to the code?
> 
> [259e37c](https://github.com/JabRef/jabref/commit/259e37c7372e5067821fdad5125a27c0e572fab0)

25d01db, I did reset and decided to keep history clean to view without rebase, so.

I will send you actual in few minutes from PC

w0nderfu11 commented:
Last one 
![Screenshot_2025-09-18-20-40-43-544_com android chrome](https://github.com/user-attachments/assets/29e6859e-f58f-41f1-8d81-9700b95f95a9)

calixtus commented:
What i meant was: What was the last commit you made real changes to the code, not style fixes?

calixtus commented:
bc i look at the commits after https://github.com/JabRef/jabref/pull/13726/commits/259e37c7372e5067821fdad5125a27c0e572fab0 and I believe I only see style fixes.

calixtus commented:
I see some meaningful changes in https://github.com/JabRef/jabref/pull/13726/commits/705c122794af2ec94cda35080058426b7ba4042b to CopyLinkedFilesAction.java, is this correct? Everything else seems to be style

w0nderfu11 commented:
25d01db
This is last one, here I got your main (already from previous commits) and fix few tests and style on 3 files.

w0nderfu11 commented:
one sec, i opened my laptop, i will tell you is https://github.com/JabRef/jabref/commit/705c122794af2ec94cda35080058426b7ba4042b OK or no

w0nderfu11 commented:
> e.g. this should not have happend in commit [da6ef2e](https://github.com/JabRef/jabref/commit/da6ef2ebf2fde10995f604038c27e0c8df17a49c) with proper style installed.
> 
> <img alt="grafik" width="852" height="179" src="https://private-user-images.githubusercontent.com/50491877/491221510-47bf176c-e11d-4567-b63f-1c0e25006791.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgyMjIxMjgsIm5iZiI6MTc1ODIyMTgyOCwicGF0aCI6Ii81MDQ5MTg3Ny80OTEyMjE1MTAtNDdiZjE3NmMtZTExZC00NTY3LWI2M2YtMWMwZTI1MDA2NzkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTE4VDE4NTcwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3MGI5ZGUzMDgyOWQwZTU5YTU4ZjY5NjI4NWVjYTVhMGRhNGM2MzhmYjE4ZDVlM2JkNjQ5YWZhZTVkOTZlMWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.jzWa7Uv3TtcVCsDN7WZ7H2gEXcbF6rUZ4ul2bpBdmEk">

this commit was reseted back as i said (idk do you see it in history)

w0nderfu11 commented:
> What i meant was: What was the last commit you made real changes to the code, not style fixes?

Here is a correct order for non style fixes (both, first one - logic, after few fixes for style etc.)
[705c122](https://github.com/JabRef/jabref/pull/13726/commits/705c122794af2ec94cda35080058426b7ba4042b)

[25d01db](https://github.com/JabRef/jabref/pull/13726/commits/25d01db37f684e279061ac07ee7a92ed2b60d28b)

jabref-machine commented:
Hey, we noticed that you **force-pushed** your changes. Force pushing is a bad practice when working together on a project (mainly because it is [not supported well by GitHub itself](https://github.com/orgs/community/discussions/3478)). Commits are lost and comments on commits lose their context, thus making it harder to review changes.

When the pull request is getting integrated into `main`, all commits will be [squashed](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits) anyway. Thus, your individual commit history will not be visible in `main`.

In future, **please avoid that**. For now, you can continue working.


<!-- ghprcomment
PR Tests
no-force-push
-->

calixtus commented:
Had to forcepush to undo the defect commits. Files changed back to 20 again. Please check if your code changes are still in and fine. If you give your ok, we will do our reviews.

w0nderfu11 commented:
Omg, I got why but bad :(
I will check tomorrow already, have a good night

calixtus commented:
Don't worry about the failing tests in this CI run, there seems to be a problem with github caches and our CI setup. We will look into that issue asap - which could be a little bit (got to worry about our day jobs too 😅 )

w0nderfu11 commented:
@calixtus hi, so for now what i should do, only review is my feature up to date in force-push commit, right?
<img width="1104" height="233" alt="image" src="https://github.com/user-attachments/assets/bcc25c51-a10f-4900-b99f-4e38ed43eb09" />

w0nderfu11 commented:
@calixtus i merged this branch from PR to my copy (test bracnh to see diff), there are no changes i done in last 2 commits.

There are no even few old fixes (major changes, asd :( )

Do i have too reupdate classes and do push again or how we will do?

For example: i renamed key for Open folder(s), there are no even this changes, unfortunately

<img width="1877" height="978" alt="image" src="https://github.com/user-attachments/assets/f93d4184-5e9d-47f8-bd41-f4e0ff68e6ac" />

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

jabref-machine commented:
Your code currently does not meet [JabRef's code guidelines](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). We use [Checkstyle](https://checkstyle.sourceforge.io/) to identify issues. You can see which checks are failing by locating the box "Some checks were not successful" on the pull request page. To see the test output, locate "Source Code Tests / Checkstyle (pull_request)" and click on it.

In case of issues with the import order, double check that you [activated Auto Import](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#enable-proper-import-cleanup). You can trigger fixing imports by pressing <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> to trigger [Optimize Imports](https://www.jetbrains.com/guide/tips/optimize-imports/).

Please carefully follow [the setup guide for the codestyle](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html). Afterwards, please [run checkstyle locally](https://devdocs.jabref.org/getting-into-the-code/guidelines-for-setting-up-a-local-workspace/intellij-13-code-style.html#run-checkstyle) and fix the issues, commit, and push.


<!-- ghprcomment
Source Code Tests
Checkstyle
-->

jabref-machine commented:
Your pull request modified git submodules.

Please follow our [FAQ on submodules](https://devdocs.jabref.org/code-howtos/faq.html#submodules) to fix.


<!-- ghprcomment
PR Tests
Submodules not modified
-->

jabref-machine commented:
Your pull request modified git submodules.

Please follow our [FAQ on submodules](https://devdocs.jabref.org/code-howtos/faq.html#submodules) to fix.


<!-- ghprcomment
PR Tests
Submodules not modified
-->

w0nderfu11 commented:
If all is ok you can get my last commit correctly, if you will found some few thing to fix i will try to do it fast, wanna make this issue done already :)

jabref-machine commented:
Your pull request modified git submodules.

Please follow our [FAQ on submodules](https://devdocs.jabref.org/code-howtos/faq.html#submodules) to fix.


<!-- ghprcomment
PR Tests
Submodules not modified
-->

w0nderfu11 commented:
Sorry if what :(

calixtus commented:
Ok, the only change between https://github.com/JabRef/jabref/pull/13726/commits/e3eb1dfdb7db884ea86ff3d13d42c4702bee8b14 and https://github.com/JabRef/jabref/pull/13726/commits/91bc19affcc1b7bd740689ed540e28d9b220f794 seems to be here: https://github.com/JabRef/jabref/pull/13726/files/514232cd3e87fd4f51e05b9264ab04674326be46..91bc19affcc1b7bd740689ed540e28d9b220f794

I think we can make this ready for review now and we will look into the PR in the next days for final review. Thanks for you endurance so far.

w0nderfu11 commented:
> I think we can make this ready for review now and we will look into the PR in the next days for final review. Thanks for you endurance so far.

Thank you too, i tried to do it clean but anyway got some fails)

W8iting for you review info, thx

w0nderfu11 commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/MultiSelectionMenuBuilderTest.java:
```diff
@@ -0,0 +1,304 @@
+        );
+    }
+
+    private static ObservableOptionalValue<BibEntry> mockEmptyBibEntryOptional() {
+        @SuppressWarnings("unchecked")
+        ObservableOptionalValue<BibEntry> optional =
+                (ObservableOptionalValue<BibEntry>) mock(ObservableOptionalValue.class);
+        when(optional.getValue()).thenReturn(Optional.empty());
+        return optional;
+    }
```
I changed for Java implementation. Before i had problems with it, ty

w0nderfu11 commented:
I will fix submodules after all reviews will be done to not make noise

calixtus commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/MultiSelectionMenuBuilderTest.java:
```diff
@@ -0,0 +1,308 @@
+        LinkedFileViewModel onlineLink2 = mockOnlineLinkViewModel("https://host/c.pdf");
+        ObservableList<LinkedFileViewModel> selectionB = FXCollections.observableArrayList(onlineLink1, onlineLink2);
+        List<MenuItem> itemsB = builder.buildMenu(selectionB);
+        assertTrue(itemsB.get(1).isDisable(), "OPEN_FOLDER should be disabled when there are no local existing files");
```
ignore

calixtus commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/MultiSelectionMenuBuilderTest.java:
```diff
@@ -0,0 +1,308 @@
+        items.get(5).fire();
+
+        Path copied = exportDir.resolve(sourceFile.getFileName());
+        assertTrue(Files.exists(copied), "Copied file must exist in export directory");
```
ignore

calixtus commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecksTest.java:
```diff
@@ -0,0 +1,166 @@
+        @Test
+        void isLocalAndExistsReturnsFalseWhenOfflineMissing() {
+            LinkedFileViewModel fileViewModel = mockOfflineMissingFileViewModel();
+            assertFalse(checks.isLocalAndExists(fileViewModel));
```
ignore

calixtus commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecksTest.java:
```diff
@@ -0,0 +1,166 @@
+
+        @Test
+        void isOnlineReturnsTrueForOnlineLink() {
+            assertTrue(checks.isOnline(mockOnlineLinkViewModel("https://x")));
```
ignore

calixtus commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecksTest.java:
```diff
@@ -0,0 +1,166 @@
+            Path path = Path.of("c.pdf");
+            LinkedFileViewModel movable = mockOfflineExistingFileViewModel(path);
+            when(movable.isGeneratedPathSameAsOriginal()).thenReturn(false);
+            assertTrue(checks.isMovableToDefaultDir(movable));
```
ignore

calixtus commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecksTest.java:
```diff
@@ -0,0 +1,166 @@
+            Path path = Path.of("c.pdf");
+            LinkedFileViewModel samePath = mockOfflineExistingFileViewModel(path);
+            when(samePath.isGeneratedPathSameAsOriginal()).thenReturn(true);
+            assertFalse(checks.isMovableToDefaultDir(samePath));
```
ignore

calixtus commented:
I just made a quick commit about using JSpecify annotations, since we are in the middle of transition to compiler time null safety.

Go on, make your changes.

calixtus commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/MultiSelectionMenuBuilder.java:
```diff
@@ -0,0 +1,186 @@
+                if (exportDirectory.isEmpty()) {
+                    return;
+                }
+                Path exportPath = exportDirectory.get();
```
Ignore

w0nderfu11 commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/MultiSelectionMenuBuilderTest.java:
```diff
@@ -0,0 +1,308 @@
+    @Test
+    void shouldSupportOnlyWhenSelectionHasMoreThanOneItem() {
+        MultiSelectionMenuBuilder builder = newBuilder(mock(LinkedFilesEditorViewModel.class), mock(DialogService.class));
+        assertFalse(builder.supports(FXCollections.observableArrayList()), "Empty selection should not be supported");
```
Fixed

w0nderfu11 commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/MultiSelectionMenuBuilderTest.java:
```diff
@@ -0,0 +1,308 @@
+        LinkedFileViewModel onlineLink = mockOnlineLinkViewModel("https://host/a.pdf");
+        ObservableList<LinkedFileViewModel> selectionA = FXCollections.observableArrayList(localExisting, onlineLink);
+        List<MenuItem> itemsA = builder.buildMenu(selectionA);
+        assertFalse(itemsA.get(1).isDisable(), "OPEN_FOLDER should be enabled when at least one local existing file is present");
```
Fixed

w0nderfu11 commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecksTest.java:
```diff
@@ -0,0 +1,166 @@
+
+        @Test
+        void isOnlineReturnsFalseForOfflineFile() {
+            assertFalse(checks.isOnline(mockOfflineExistingFileViewModel(Path.of("b.pdf"))));
```
@calixtus even IDE tells i can simplify equals to false (assert), i did it with the same mindset but i changed to assertEquals in all cases above.

Fixed

calixtus commented:
take tragbots comments as suggestions, not as requirements. If you have reasons not follow tragbot, comment under it.
Tragbot is not always right.

calixtus commented on jabgui/src/test/java/org/jabref/gui/fieldeditors/contextmenu/ContextMenuFactoryTest.java:
```diff
@@ -1,166 +1,182 @@
 
-        return file;
+        assertNotNull(contextMenu);
+        assertTrue(contextMenu.getItems().isEmpty(), "Menu should be empty for empty selection");
```
No. Use what junit offers.

subhramit commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/ContextAction.java:
```diff
@@ -29,40 +38,70 @@ public ContextAction(StandardActions command,
 
+        Observable entryFieldsObservable = bibEntry.getValue()
+                                                   .map(BibEntry::getFieldsObservable)
+                                                   .orElse(EMPTY_FIELDS);
```
context: constant declared but used only here

subhramit replied:
again, this is a nitpick so you can ignore this as well for this PR

w0nderfu11 commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditorViewModel.java:
```diff
@@ -104,6 +110,39 @@ public static LinkedFile fromFile(Path file, List<Path> fileDirectories, Externa
+            anyDownloadOngoing.set(
+                    files.stream().anyMatch(fileViewModel -> fileViewModel.downloadOngoingProperty().get())
+            );
+            maxDownloadProgress.set(
+                    files.stream()
+                         .mapToDouble(fileViewModel -> fileViewModel.downloadProgressProperty().get())
+                         .max()
+                         .orElse(0.0)
+            );
+        };
```
@subhramit agree, will change it, i missed it will show 1.0 if only 1 file from group was downloaded

w0nderfu11 replied:
@subhramit i changed for average and logic works broken, status that files are downloaded is not correct( they are downloaded, but the name in menu, status and additional commands in menu don`t know about it), i can fix it but again i have to add bunch of code, i can simply do back to .max(), is it ok?

I mean this worked correct with this function, can i do it or is it principal to use .average() for readability? I know it would be better without .max() but as i said this PR is long enough is in progress so i prefer to do it back

subhramit commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditor.java:
```diff
@@ -301,36 +310,31 @@ private void addFromURL() {
     }
 
     private void handleItemMouseClick(LinkedFileViewModel linkedFile, MouseEvent event) {
-        if (event.getButton() == MouseButton.PRIMARY && (event.getClickCount() == 2)) {
-            linkedFile.open(); // Double-click: open file
-        } else if (activeContextMenu != null && event.getButton() == MouseButton.PRIMARY) {
-            activeContextMenu.hide(); // Hide context menu if left-click
-            activeContextMenu = null;
+        if (event.getButton() == MouseButton.PRIMARY) {
+            if (event.getClickCount() == 2) {
+                linkedFile.open();
+                event.consume();
+                return;
+            }
+            if (activeContextMenu != null) {
+                activeContextMenu.hide();
+                activeContextMenu = null;
+            }
         } else if (event.getButton() == MouseButton.SECONDARY) {
             if (activeContextMenu != null) {
-                activeContextMenu.hide(); // Hide any existing context menu
+                activeContextMenu.hide();
```
do not touch existing code unless you need to (in which case you should justify and take approval for the reason before going forward)

subhramit commented:
@w0nderfu11 after the changes are made, can you please put a video recording of your changes in action with your demo linked files? Once we see it working, it will greatly reduce the friction and time to merge your PR, given this was a lengthy issue with lots of changes and follow ups (eg whether to use max or average for download of linked files). You can put all changes in a single recording or different ones (and label them accordingly).
You can use Loom or OBS studio for recording.

w0nderfu11 commented:
> @w0nderfu11 after the changes are made, can you please put a video recording of your changes in action with your demo linked files? Once we see it working, it will greatly reduce the friction and time to merge your PR, given this was a lengthy issue with lots of changes and follow ups (eg whether to use max or average for download of linked files). You can put all changes in a single recording or different ones (and label them accordingly). You can use Loom or OBS studio for recording.

@subhramit sure, i tested whole functionality in app many times, now all is working, i will do fixes you said above (interface ---> class, max ---> average and delete non prift constant) and record you a video of multi-context menu in action 😃 

I hope i got you right, you meant video in GUI (application running)?

calixtus commented:
Comment in github about tragbot, not in code

subhramit commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecks.java:
```diff
@@ -0,0 +1,69 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static java.util.Objects.requireNonNull;
```
1. Use jspecify annotations
2. Don't import static.

w0nderfu11 replied:
Will fix, ty

w0nderfu11 replied:
Fixed

subhramit commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/MultiSelectionMenuBuilder.java:
```diff
@@ -56,34 +56,36 @@ public List<MenuItem> buildMenu(ObservableList<LinkedFileViewModel> selection) {
-        menuItems.add(buildCopyToFolderItem(actionFactory, selection));
+        SelectionChecks checks = new SelectionChecks(databaseContext, preferences);
+
+        menuItems.add(batchCommandItem(actionFactory, StandardActions.OPEN_FILE, selection, _ -> true));
```
Can you explain this change to me? You changed `this::isLocalAndExists` to `_ -> true`??

w0nderfu11 replied:
@subhramit it worked that for selected files (only LOCAL!) it was working correctly, i deleted predicate cause we can open all files, URL files and Local files, so now it works correctly

calixtus replied:
ok, for the sake of clean and readable, code maybe just add a dummy method `alwaysEnabled` to SelectionChecks.

subhramit commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecks.java:
```diff
@@ -15,42 +15,50 @@
+import static java.util.Objects.requireNonNull;
 
-    Logger LOGGER = LoggerFactory.getLogger(SelectionChecks.class);
+public record SelectionChecks(BibDatabaseContext databaseContext, GuiPreferences preferences) {
```
What was the rationale behind converting this to a record?
You are still using the methods here in the multiselectionmenubuilder class only. Why this additional layer of abstraction?

w0nderfu11 replied:
@subhramit

I CAN MERGE theese methods in multiselection menu builder if you want, i guess it will reduce code and abstraction as you said. IT was planned to use in other class for Single too but now there is no need.

subhramit commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditor.java:
```diff
@@ -253,13 +266,9 @@ private void setUpKeyBindings() {
         listView.addEventFilter(KeyEvent.KEY_PRESSED, event -> {
             Optional<KeyBinding> keyBinding = preferences.getKeyBindingRepository().mapToKeyBinding(event);
             if (keyBinding.isPresent()) {
-                switch (keyBinding.get()) {
-                    case DELETE_ENTRY:
-                        deleteAttachedFilesWithConfirmation();
-                        event.consume();
-                        break;
-                    default:
-                        // Pass other keys to children
+                if (keyBinding.get() == KeyBinding.DELETE_ENTRY) {
+                    deleteAttachedFilesWithConfirmation();
+                    event.consume();
```
Why was the default case handling not needed anymore?

calixtus replied:
We kept this deliberatly a switch to keep it consistent with keybindings check in other ui components.

w0nderfu11 commented:
> > @subhramit @calixtus i changed interface to class, fixed few moments (cause i did test inside GUI really long time ago), deleted variable we discussed, i let .max() to be cause all logic works correctly. Video is too big, so i prepared google drive link for you:
> > https://drive.google.com/file/d/14NRhxLdCbD-H9YNAt7p7ycETdTHrQBS_/view?usp=sharing
> > I had lags (caused not by JabRef app), so some moments caused by it (you will see in video)
> > P.S. bot now suggest to switch for assertTrue/False, but ihad it before, funny, i will ignore it if you don`t mind
> > in my point of view all is ok and usable enough, logic is correct, i tried to optimize all i could (i mean you Set for folders etc.)
> 
> 2:32 onwards, when doing "copy linked file(s) to folder", only one file is copied out of all selected. Why?

@subhramit this is case with 1 online file and 1 local file, copied only 1 local file, guess this is all explanation

i mean the online file is not downloaded

subhramit commented:
> > > @subhramit @calixtus i changed interface to class, fixed few moments (cause i did test inside GUI really long time ago), deleted variable we discussed, i let .max() to be cause all logic works correctly. Video is too big, so i prepared google drive link for you:
> > > https://drive.google.com/file/d/14NRhxLdCbD-H9YNAt7p7ycETdTHrQBS_/view?usp=sharing
> > > I had lags (caused not by JabRef app), so some moments caused by it (you will see in video)
> > > P.S. bot now suggest to switch for assertTrue/False, but ihad it before, funny, i will ignore it if you don`t mind
> > > in my point of view all is ok and usable enough, logic is correct, i tried to optimize all i could (i mean you Set for folders etc.)
> > 
> > 
> > 2:32 onwards, when doing "copy linked file(s) to folder", only one file is copied out of all selected. Why?
> 
> @subhramit this is case with 1 online file and 1 local file, copied only 1 local file, guess this is all explanation
> 
> i mean the online file is not downloaded
> 
> I need to change only things above right? Don`t wanna spam with changes so let`s make a deal what changes we need

In the issue, the requirement comment says:
> All files are copied to the selected destination folder

So this is wrong/misleading. The option doesn't say "Copy downloaded file(s) to to folder".

calixtus commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/MultiSelectionMenuBuilder.java:
```diff
@@ -0,0 +1,188 @@
+                        continue;
+                    }
+                    Path source = sourcePath.get();
+                    Path target = exportPath.resolve(source.getFileName());
```
ignore this. Strange, that trag-bot read this as Paths.get

calixtus commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecks.java:
```diff
@@ -0,0 +1,69 @@
+package org.jabref.gui.fieldeditors.contextmenu;
+
+import java.io.IOException;
+import java.nio.file.Path;
```
Ignore this.

calixtus commented on jabgui/src/main/java/org/jabref/gui/fieldeditors/contextmenu/SelectionChecks.java:
```diff
@@ -0,0 +1,69 @@
+
+        for (Path fileToSelect : representativeByDir.values()) {
+            try {
+                NativeDesktop.get().openFolderAndSelectFile(fileToSelect);
```
Ignore this. trag-bot is silly sometimes.

w0nderfu11 commented:
Hi, @calixtus let`s make a summary and i will start work - so i fix thing above, and as @koppor said, split logic to package *logic*, right?

w0nderfu11 commented:
@koppor @calixtus @subhramit about merging to package ...logic. Would you mind if i create package in org,jabref.logic ---> or.jabref.logic.files

Is it ok or i have to use existing package? I reviewed them and i can say create new one would be good.

Other way - i found LinkedFileHandler in package org.jabref.logic.externalfiles

What do you prefer?

<img width="757" height="552" alt="image" src="https://github.com/user-attachments/assets/155fcf36-ab2b-4936-a2a1-552471a75ad9" />

calixtus commented:
Hey, we briefly talked about the refactoring, @koppor suggested. We think we should do this in a follow-up. So forget about moving stuff to logic and just polish this one. Then we finalize and merge.
